llvm/lib/Target/AIE/AIEMaxLatencyFinder.cpp

```diff
@@ -221,17 +221,25 @@ unsigned MaxLatencyFinder::operator()(MachineInstr &MI) {
     return Latency;
   }
 
-  LLVM_DEBUG(dbgs() << "Earliest for: " << MI);
-  // Track the earliest use in any successor block, given the cycles in
-  // which these uses are scheduled
+  const BlockState &BS = IB.getBlockState(CurBB);
+
+  // Compute the earliest bottom-up cycle in which MI can be scheduled given
+  // known schedule successor instructions. We start conservative, and then
+  // reduce that number after looking into those instructions.
+  // Note that the scheduled instructions might be in the same region, so we
+  // need to account for them.
+  const int NumScheduledInstrsInRegion = BS.BottomInsert.size();
   int Earliest = Latency;
+
+  LLVM_DEBUG(dbgs() << "Earliest for: " << MI);
   for (ArrayRef<MachineBundle> TopBundles :
        getScheduledSuccessorInstructions(*CurBB, *Scheduler)) {
-    Earliest = findEarliestRef(MI, TopBundles, Earliest);
+    Earliest =
+        findEarliestRef(MI, TopBundles, Earliest + NumScheduledInstrsInRegion);
   }
 
   LLVM_DEBUG(dbgs() << "   Earliest=" << Earliest << "\n");
-  Latency = std::max(Latency - Earliest, 1);
+  Latency = std::max(Latency + NumScheduledInstrsInRegion - Earliest, 1);
```

Comment:
I'm not the biggest fan of this change in `AIEMaxLatencyFinder`. It would probably be better to extend the `EmitFixedSUnits` so that it also adds edges between "free" and "fixed" instructions.